### ISSUE TITLE
Sometimes, editing the latest 200 tokens of text -> resamples all 3000 tokens of context.

### ISSUE THREAD

**Issue by WingFoxie** [LostRuins/koboldcpp#614]
Sadly, although this problem bothered me since forever, I still haven't find a clear pattern for how this bug triggers.
(Definition of the "whole context": All text sampled within the latest story. Doesn't include the "Memory" section. That section never gets resampled unless I edit the "Memory" directly.)
(For example: My "Max Ctx. Tokens" is set to 4096, but there are about 1000 tokens of "Memory". That leaves me 3000 tokens of text getting resampled when the bug happens.)

(**Use "Story Mode" if you want to test any of there below.** I mean no "end sequences" whatsoever, so the AI will very likely generate as much as it could.)

What's most likely to trigger a huge resample (At least try this one please.):
(How to test this: Load a story that is at least a bit longer than the context length, click "submit" once to make sure all the texts are sampled. And use Story mode.)
**"Amount to Gen."=50.** Generate once, generate twice, 
delete the whole 100 tokens of output, 
submit again -> resampling the whole context. 
(If that doesn't work for you, try **generating 4 times in a row (200 tokens) before deleting them all and submit again**.)

Why I have to do this, instead of editing and finishing the text 50 tokens at a time:
Occasionally AI says something that's much longer than 50 tokens, and I'm not sure whether to keep the text or not. So I ended up generating a couple of times in a row, before deleting them all. Which triggers huge resampling. 
But I still only deleted like only the latest 200 tokens of text of so! Way less than 3000 tokens.
And most of the time 50 tokens is enough for the output drift to the point that an edit is required, so I won't increase that from 50.

<details>
<summary>Other less important cases I considered:</summary>
<p>

What's more likely to (but not always will) trigger a huge resample:
Complex edits. Edit some words of a sentence without removing the whole thing.
Remove all texts of one generation completely.
Edit stuff that's generated a while ago, like 300 tokens ago. (One screen of text is about 300 tokens on my configuration, so I'm still editing stuff on the latest screen.)

What's not likely to trigger a huge resample:
Simple edits. Deleting the last 30 tokens, for example.
Don't remove a whole generation, keep at least one letter or something.

What never triggers a huge resample:
Only editing the latest response, all the time.
Moving the old, out of context text out of the text box, and submit again.
</p>
</details>

<details>
<summary>Reality check:</summary>
<p>

Yes I definitely was keeping enough text in the text box. Like maybe 5000 tokens. So there's always enough room for "context shifting" to go back.
And I was chatting for a while, so all the former tokens were sampled at some point in the same run.
And I keep the "context shifting" feature at its default setting, which is "on". I'm sure that I'm using "context shifting" not "smart context".
And in my configuration when I keep clicking "generate more" it will keep generating for sure. No instant EOS or loop... (Was using other models that will refuse to generate even though I clearly left if with half a sentence to complete.)
When I have to eventually move some very old text away, I can avoid a huge resampling, by checking the "Last Request", and make sure that I didn't move away any text inside that range.
Mostly using default settings, except changing the "Repetition Penalty" from "1" to "1.1".

Oh, and, yes, the resampling happens too frequently to be considered a bug. If a full resample only happens with every 4000 new tokens or so. I wouldn't complain. But it's obviously not the case I'm experiencing. I can keep triggering resampling every 100 tokens or so, if I really try.
</p>
</details>

Plus: Sometimes when I edit the text like 200 tokens ago. The program will fail and succeed at the same time.
It will only resample like the latest 200 tokens of text, but then refuse to generate more, says that there's not enough context buffer or something... 
Which is clearly wrong, clicking the submit button again and it will generate just fine. 
(So I just **have to click the submit button, one more time than I should**, in this case.)

Oh and why do I resort to using "Story Mode"??
Back then it was a User and AI chat, I was using "Chat Mode". I sometimes will write half the AI's response, and let the program generate the second half.
But now it is a User and AI1 and AI2 chat. The program, in "Chat Mode", **keeps refusing to finish the half a response for AI2!** It only works for AI1! This is unbearable and causes be to resort to using "Story Mode" altogether. I can get used to the user names not colored anymore... I mean I have to.

The bottom line is, no matter how I edit, I'm only editing or deleting the text within the last 300 tokens of the 3000 latest sampled story text. So under no (at least almost no) circumstances should the whole 3000 tokens be sampled, right?
